### ISSUE TITLE
From user study: No useful error message shows up when creating a block blob client to download a blob with a blob name that doesn't exist

### ISSUE THREAD

**Issue by ahsonkhan** [Azure/azure-sdk-for-cpp#4689]
The expectation is that whenever there is an error in the developer code (for example something like a request failed exception), there is some information printed to the console that can help provide a clue on what went wrong.

Following the instructions from here, the user ends up with the following code:
https://learn.microsoft.com/en-us/azure/storage/blobs/quickstart-blobs-c-plus-plus?tabs=environment-variable-windows#upload-blobs-to-a-container

In the study, there is no blob in the container with the name `blob.txt`, instead, it is called `blob1.txt.` When the user runs their application (which has no compiler error), it doesn't show any information of why it failed.

```C++
const string storageConnectionString = getenv("AZURE_STORAGE_CONNECTION_STRING");
const string containerName = "blobcontainer";

// TASK 1: Create a Storage Blob Container Client
cout << "Creating Storage Blob Container Client" << endl;

// Initialize a new instance of BlobContainerClient
BlobContainerClient containerClient
    = BlobContainerClient::CreateFromConnectionString(storageConnectionString, containerName);

// Create the container. This will do nothing if the container already exists.
std::cout << "Creating container: " << containerName << std::endl;
containerClient.CreateIfNotExists();

// TASK 2: List the Blobs in the Storage Container
cout << "Listing blobs..." << endl;
auto listBlobsResponse = containerClient.ListBlobs();
for (auto blobItem : listBlobsResponse.Blobs)
{
    std::cout << "Blob name: " << blobItem.Name << std::endl;
}

// TASK 3: Download a Blob from the Storage Container
cout << "Downloading Blob" << endl;

// ISSUE: There is no blob in the container with this name
std::string blobName = "blob.txt";

// ISSUE: Should this fail with a visible error message for a clue of what went wrong? Maybe.
BlockBlobClient blobClient = containerClient.GetBlockBlobClient(blobName);
auto properties = blobClient.GetProperties().Value;
std::vector<uint8_t> downloadedBlob(properties.BlobSize);

// ISSUE: This should certainly fail with an appropriate error message, which highlights something
// like an RequestFailedException outputted to the console.
blobClient.DownloadTo(downloadedBlob.data(), downloadedBlob.size());
std::cout << "Downloaded blob contents: " << std::string(downloadedBlob.begin(), downloadedBlob.end()) << std::endl;
```

Sample output which doesn't show what went wrong:
```text
Creating Storage Blob Container Client
Creating container: blobcontainer
Listing blobs...
Blob name: blob1.txt
Blob name: blob2.txt
Blob name: blob3.txt
Downloading Blob

C:\Users\CppDev\source\repos\CppStudySample\out\build\x64-release\CppStudySample.exe (process 11212) exited with code -1073740791.
```

When the user runs such an app, there is no error message printed.

cc @ronniegeraghty, @Jinming-Hu

**Comment by ahsonkhan** [comment]
The appropriate exception, reason phrase, and call stack shows up when debugging the application with a breakpoint on `blobClient.GetProperties()`, but the exception doesn't seem to be thrown when running in release and the information doesn't end up on the console:
![image](https://github.com/Azure/azure-sdk-for-cpp/assets/6527137/7dc7ff1c-7cce-4e8c-a187-57f6d2737d04)

**Comment by Jinming-Hu** [comment]
If the customer had wrapped the code with a `try {...} catch` block, he should've see the error message.

I think this is more of a question of capturing exception and printing error message from there. The customer cannot figure out the problem if he doesn't catch exception, no matter which service he's using or the request fails due to whatever reason

**Comment by LarryOsterman** [comment]
> If the customer had wrapped the code with a `try {...} catch` block, he should've see the error message.
> 
> I think this is more of a question of capturing exception and printing error message from there. The customer cannot figure out the problem if he doesn't catch exception, no matter which service he's using or the request fails due to whatever reason

I've seen a similar set of issues with the AMQP samples. 

IMHO, because samples are supposed to express optimal customer behavior, they should have a try/catch wrapped around their main which outputs the `.what()` to `cerr`.  This is especially important on Windows where the unhandled exception behavior is ... suboptimal.

**Comment by ahsonkhan** [comment]
I agree that we'd want to highlight best practices in our samples, which include try/catch and error handling, since customers certainly start off with copy-paste/modify, and it becomes a great teaching tool (cc @ronniegeraghty as part of docs/samples improvements).

I haven't tried yet, @LarryOsterman have you observed the unhandled exception behavior on linux be better, where we see the contents of the exception, unlike Windows?

**Comment by LarryOsterman** [comment]
> I agree that we'd want to highlight best practices in our samples, which include try/catch and error handling, since customers certainly start off with copy-paste/modify, and it becomes a great teaching tool (cc @ronniegeraghty as part of docs/samples improvements).
> 
> I haven't tried yet, @LarryOsterman have you observed the unhandled exception behavior on linux be better, where we see the contents of the exception, unlike Windows?

If my memory serves me correctly, on Linux, we get a diagnostic like "unhandled exception thrown: " followed by ex.what(). On Windows we get a MessageBox with a comment about the application terminating unexpectedly.
